File: Core/src/main/java/com/hqq/core/permission/PermissionsHasImpl.java
package com.hqq.core.permission;

import android.Manifest;
import android.content.Context;

import java.util.Arrays;

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.permission
 * @FileName :   PermissionsHasImpl
 * @Date : 2018/11/22 0022  上午 9:57
 * @Email :  qiqiang213@gmail.com
 * @Descrive : 判断是否拥有权限
 */
public class PermissionsHasImpl implements IPermissionsHas {

    /**
     * 权限存在依赖关系  如定位权限 需要读写权限
     * 用吃方法判断
     *
     * @param context
     * @param permissions
     * @return
     */
    @Override
    public boolean hasPermission(Context context, String... permissions) {
        boolean result = true;
        if (permissions != null && permissions.length > 0) {
            for (String permission : permissions) {
                if (Arrays.asList(IPermissionsHas.getCalendar()).contains(permission)) {
                    result = result && hasCalendar(context);
                } else if (Arrays.asList(IPermissionsHas.getCamera()).contains(permission)) {
                    result = result && hasCamera(context);
                } else if (Arrays.asList(IPermissionsHas.getContacts()).contains(permission)) {
                    result = result && hasContacts(context);
                } else if (Arrays.asList(IPermissionsHas.getLocation()).contains(permission)) {
                    result = result && hasLocation(context);
                } else if (Arrays.asList(IPermissionsHas.getMicrophone()).contains(permission)) {
                    result = result && hasMicrophone(context);
                } else if (Arrays.asList(IPermissionsHas.getPhone()).contains(permission)) {
                    result = result && hasPhone(context);
                } else if (Arrays.asList(IPermissionsHas.getSensors()).contains(permission)) {
                    result = result && hasSensors(context);
                } else if (Arrays.asList(IPermissionsHas.getSMS()).contains(permission)) {
                    result = result && hasSMS(context);
                } else if (Arrays.asList(IPermissionsHas.getStorage()).contains(permission)) {
                    result = result && hasStorage(context);
                }
            }
        }
        return result;
    }

    @Override
    public boolean hasCalendar(Context context) {
        //(允许程序读取用户的日程信息)
        return IPermissionActions.hasPermission(context,
                IPermissionsHas.getCalendar()
        );

    }

    @Override
    public boolean hasCamera(Context context) {
        //(允许访问摄像头进行拍照)
        return IPermissionActions.hasPermission(context,
                IPermissionsHas.getCamera());
    }

    @Override
    public boolean hasContacts(Context context) {
        return hasStorage(context) && IPermissionActions.hasPermission(context,
                IPermissionsHas.getContacts()
        );
    }

    @Override
    public boolean hasLocation(Context context) {
        return hasStorage(context) && IPermissionActions.hasPermission(context, IPermissionsHas.getLocation());

    }

    @Override
    public boolean hasMicrophone(Context context) {
        return IPermissionActions.hasPermission(context, IPermissionsHas.getMicrophone());

    }

    @Override
    public boolean hasPhone(Context context) {
        return IPermissionActions.hasPermission(context, IPermissionsHas.getPhone());
    }

    @Override
    public boolean hasSensors(Context context) {
        return IPermissionActions.hasPermission(context, IPermissionsHas.getSensors());
    }

    @Override
    public boolean hasSMS(Context context) {
        return IPermissionActions.hasPermission(context,
                IPermissionsHas.getSMS()
        );
    }

    @Override
    public boolean hasStorage(Context context) {
        return IPermissionActions.hasPermission(context, IPermissionsHas.getStorage());
    }
}

File: Core/src/main/java/com/hqq/core/permission/PermissionsHasImpl.kt
package com.hqq.core.permission

import android.content.Context
import java.util.*

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.permission
 * @FileName :   PermissionsHasImpl
 * @Date : 2018/11/22 0022  上午 9:57
 * @Email :  qiqiang213@gmail.com
 * @Descrive : 判断是否拥有权限
 */
class PermissionsHasImpl : IPermissionsHas {
    /**
     * 权限存在依赖关系  如定位权限 需要读写权限
     * 用吃方法判断
     *
     * @param context
     * @param permissions
     * @return
     */
    override fun hasPermission(context: Context?, vararg permissions: String): Boolean {
        var result = true
        if (permissions.isNotEmpty()) {
            for (permission in permissions) {
                if (IPermissionsHas.calendar.contains(permission)) {
                    result = result && hasCalendar(context)
                } else if (IPermissionsHas.camera.contains(permission)) {
                    result = result && hasCamera(context)
                } else if (IPermissionsHas.contacts.contains(permission)) {
                    result = result && hasContacts(context)
                } else if (IPermissionsHas.location.contains(permission)) {
                    result = result && hasLocation(context)
                } else if (IPermissionsHas.microphone.contains(permission)) {
                    result = result && hasMicrophone(context)
                } else if (IPermissionsHas.phone.contains(permission)) {
                    result = result && hasPhone(context)
                } else if (IPermissionsHas.sensors.contains(permission)) {
                    result = result && hasSensors(context)
                } else if (IPermissionsHas.sMS.contains(permission)) {
                    result = result && hasSMS(context)
                } else if (IPermissionsHas.storage.contains(permission)) {
                    result = result && hasStorage(context)
                } else if (IPermissionsHas.bluetooth.contains(permission)) {
                    result = result && hasBluetooth(context)
                }
            }
        }
        return result
    }

    private fun hasBluetooth(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.bluetooth)
    }

    override fun hasCalendar(context: Context?): Boolean {
        //(允许程序读取用户的日程信息)
        return IPermissionActions.hasPermission(context, *IPermissionsHas.calendar)
    }

    override fun hasCamera(context: Context?): Boolean {
        //(允许访问摄像头进行拍照)
        return IPermissionActions.hasPermission(context, *IPermissionsHas.camera)
    }

    override fun hasContacts(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.contacts)
    }

    override fun hasLocation(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.location)
    }

    override fun hasMicrophone(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.microphone)
    }

    override fun hasPhone(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.phone)
    }

    override fun hasSensors(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.sensors)
    }

    override fun hasSMS(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.sMS)
    }

    override fun hasStorage(context: Context?): Boolean {
        return IPermissionActions.hasPermission(context, *IPermissionsHas.storage)
    }
}